flips/20211129-cadence-mutability-restrictions.md

```diff
@@ -0,0 +1,418 @@
+
+pub fun addToY(foo: Foo, i: Int) {
+    foo.y.append(i)
+}
+```
+
+This is designed to prevent external code from mutating the values of fields it can read from your contract. Consumers
+of your contract may read the values in a `pub let` or `pub var` field, but cannot change them in any way. 
+
+If you wish to allow other code to update or modify a field in your contract, you may expose a method 
+to do so, like in the example above with `addToX`, or you may use the `pub(set)` access mode, which 
+allows any code to mutate or write to the field it applies to. 
+
+### Implementation Details 
+
+This change adds a new error, the `ExternalMutationError`, which is raised when a field 
+is mutated outside of the context in which it was defined. The error message will also
+suggest that the user instead use a setter or modifier method for that field.
+
+Specifically, the error is reported whenever a user attempts to perform an 
+index assignment on a member that is not either declared with the `pub(set)` 
+access mode, or is defined in the current enclosing scope. This check is the
+same one performed for writing to fields, with the difference that mutation 
+is allowed on both `let` and `var` fields, while only the latter can be written to.
+
+Additionally, array and dictionary methods now track an additional bit of information
+indicating whether they mutate their receiver. Mutating methods may not be called on 
+members that are not declared with the `pub(set)` access mode, or defined in the current
+enclosing scope. 
+
+The array methods that are considered mutating are `append`, `appendAll`, `remove`,
+`insert`, `removeFirst` and `removeLast`. The dictionary methods that are considered
+mutating are `insert` and `remove`. Additionally, indexed assignment (`x[y] = e`) will
+be considered mutating for both arrays and dictionaries.  
+
+The limitations on mutation are designed to closely mirror the limitations on writes, 
+so that they can be easily explained to and understood by the user in terms of 
+language principles with which they are already familiar. Similarly, the suggested
+workaround of adding a setter or modifier method to the composite type is designed 
+to be immediately recognizable to any developer familiar with object-oriented
+design principles. 
+
+### Drawbacks
+
+This has the potential to break a number of existing contracts by restricting 
+code that was previously legal. This change would require a migration path for 
+these contracts in order for developers to update their contracts to satisfy
+Cadence's new restrictions.
+
+This also removes a small amount of expressivity from the language. Previously, 
+the `pub let` declaration was a way to create a field that could be read and
+mutated in all contexts, but not written, while `pub(set) var` created
+a field that could be read, written and mutated in all contexts. After this change,
+it would not be possible to describe a field that can be read and mutated but not set, 
```

Comment:
Are there cases where this is needed that a setter function would be inadequate?

Reply:
IMO, no, there are no specific cases where you could not accomplish the same thing with a setter function, but I felt it was worth at least mentioning, since it is a concept that can no longer be expressed.